libtenzir/builtins/operators/deduplicate.cpp

```diff
@@ -179,18 +204,68 @@ class plugin final : public operator_plugin2<deduplicate_operator> {
 public:
   auto make(invocation inv, session ctx) const
     -> failure_or<operator_ptr> override {
-    auto key = std::optional<ast::expression>{};
+    auto selector_exprs = std::vector<ast::expression>{};
+    auto expression_key = std::optional<ast::expression>{};
+    auto named_args = std::vector<ast::expression>{};
+    auto seen_named = false;
+    selector_exprs.reserve(inv.args.size());
+    named_args.reserve(inv.args.size());
+    for (auto& arg : inv.args) {
+      if (is<ast::assignment>(arg)) {
+        seen_named = true;
+        named_args.push_back(std::move(arg));
+        continue;
+      }
+      if (seen_named) {
+        diagnostic::error("positional arguments must precede named arguments")
+          .primary(arg)
+          .emit(ctx);
+        return failure::promise();
+      }
+      auto selector = ast::field_path::try_from(arg);
+      if (selector) {
+        if (selector->has_this() || selector->path().empty()) {
+          diagnostic::error("cannot deduplicate `this` explicitly")
+            .primary(*selector)
+            .emit(ctx);
+          return failure::promise();
+        }
+        if (expression_key) {
+          diagnostic::error(
+            "cannot mix field selectors with general expressions")
+            .primary(arg)
+            .emit(ctx);
+          return failure::promise();
+        }
+        selector_exprs.push_back(std::move(*selector).unwrap());
+        continue;
+      }
+      if (expression_key) {
+        diagnostic::error("expected selector").primary(arg).emit(ctx);
+        return failure::promise();
+      }
+      expression_key = std::move(arg);
+    }
     auto limit = std::optional<located<int64_t>>{};
     auto cfg = configuration{};
     auto parser = argument_parser2::operator_("deduplicate");
-    parser.positional("key", key, "any");
     parser.named("distance", cfg.distance);
```

Comment:
I would actually keep an unused `keys` here:

```suggestion
    [[maybe_unused]] auto unused_key = std::optional<ast::expression>{};
    auto parser = argument_parser2::operator_("deduplicate");
    parser.positional("key", unused_key, "any");
    parser.named("distance", cfg.distance);
```

That way the synopsis that gets printed on failure looks correct